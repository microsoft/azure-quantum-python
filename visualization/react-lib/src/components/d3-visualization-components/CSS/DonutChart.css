:root {
  --d3-colors: #1a5d8c, #8c1a5c, #aebac0, #323130;
  --chart-opacity: 0.75;
  --chart-hover-opacity: 1;
  --pad-angle: 0.01;
  --dimensions-legend: 20;
  --stroke-width: 2;
}

.title {
  font-family: "Segoe UI";
  font-style: normal;
  font-weight: 600;
  font-size: 35px;
  line-height: 47px;
  display: flex;
  align-items: center;
  text-align: center;
  color: #201f1e;
  text-anchor: middle;
}

.svg-container {
  justify-content: center;
  align-items: center;
}

.legend {
  font-size: 14px;
  font-family: Segoe UI;
  line-height: 18px;
  fill: black;
  font-weight: 600;
  font-style: normal;
}

.legend-maintext {
  font-size: 14px;
  fill: #000000;
  font-family: Segoe UI;
  line-height: 18px;
  font-weight: 400;
  font-style: normal;
}

.legend-values {
  font-size: 28px;
  font-family: Segoe UI;
  line-height: 34px;
  fill: #24272b;
  font-style: normal;
  font-weight: 600;
}

.donut-middle-title {
  font-size: 18px;
  fill: #323130;
  font-weight: 400;
  font-style: normal;
  text-anchor: middle;
  font-family: Segoe UI;
  line-height: 21px;
}

.donut-middle-text {
  font-size: 55px;
  fill: #323130;
  font-weight: 400;
  font-style: normal;
  text-anchor: middle;
  font-family: Segoe UI;
  line-height: 73px;
}
