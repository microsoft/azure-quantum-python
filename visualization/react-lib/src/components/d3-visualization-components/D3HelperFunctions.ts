/*------------------------------------
  Copyright (c) Microsoft Corporation.
  Licensed under the MIT License.
  All rights reserved.
------------------------------------ */
import * as d3 from "d3";
import * as d3Format from "d3-format";

export type LegendData = {
  title: string;
  legendTitle: string;
  value: number;
};

export interface TextStyle {
  fontFamily: string;
  fontStyle: string;
  fontWeight: string;
  fontSize: string;
  lineHeight: string;
  display: string;
  alignItems: string | null | undefined;
  textAlign: string | null | undefined;
  color: string;
  textAnchor: string | null | undefined;
}

export function drawEllipses(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  cx: number,
  cy: number,
  spaceBetween: number,
  radius: number,
  fillColor: string,
) {
  svg
    .append("circle")
    .attr("cx", cx)
    .attr("cy", cy)
    .attr("fill", fillColor)
    .attr("r", radius);

  svg
    .append("circle")
    .attr("cx", cx + spaceBetween)
    .attr("cy", cy)
    .attr("fill", fillColor)
    .attr("r", radius);

  svg
    .append("circle")
    .attr("cx", cx + spaceBetween * 2)
    .attr("cy", cy)
    .attr("fill", fillColor)
    .attr("r", radius);
}

export function drawLine(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  linePoints: number[][],
  id: string,
  strokeWidth: string,
  markerStart: string,
  markerEnd: string,
  fillColor: string,
  strokeColor: string,
  isDashed: boolean,
) {
  // Create line generator
  const lineGenerator = d3.line();
  if (isDashed) {
    svg
      .append("path")
      .attr("id", id)
      .attr("d", lineGenerator(linePoints as any))
      .attr("stroke-width", strokeWidth)
      .attr("marker-start", markerStart)
      .attr("marker-end", markerEnd)
      .style("fill", fillColor)
      .style("stroke-dasharray", "3,3")
      .style("stroke", strokeColor);
  } else {
    svg
      .append("path")
      .attr("id", id)
      .attr("d", lineGenerator(linePoints as any))
      .attr("stroke-width", strokeWidth)
      .attr("marker-start", markerStart)
      .attr("marker-end", markerEnd)
      .style("fill", fillColor)
      .style("stroke", strokeColor)
      .lower();
  }
}

export function drawLegend(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  legendData: LegendData[],
  midpoint: number,
  chartBottomY: number,
  chartStartX: number,
  legendColor: d3.ScaleOrdinal<string, unknown, never>,
  showLegendValues: boolean,
  useTitleForColor: boolean,
) {
  const legend = svg
    .selectAll(".legend")
    .data(legendData)
    .enter()
    .append("g")
    .attr(
      "transform",
      (d, i) => `translate(${midpoint * i + chartStartX}, ${chartBottomY})`,
    );

  if (useTitleForColor) {
    legend
      .append("rect")
      .attr("width", 20)
      .attr("height", 20)
      .attr("x", 0)
      .attr("y", 50)
      .style("fill", (d) => legendColor(d.title) as string);
  } else {
    legend
      .append("rect")
      .attr("width", 20)
      .attr("height", 20)
      .attr("x", 0)
      .attr("y", 50)
      .style("fill", (d) => legendColor(d.legendTitle) as string);
  }

  legend
    .append("text")
    .attr("x", 25)
    .attr("y", 60)
    .text((d) => `${d.title}`)
    .style("font-size", "14px")
    .style("font-family", "Segoe UI")
    .style("line-height", "18px")
    .style("fill", "black")
    .style("font-weight", "600")
    .style("font-style", "normal");

  legend
    .append("text")
    .attr("x", 25)
    .attr("y", 75)
    .text((d) => `${d.legendTitle}`)
    .style("font-size", "14px")
    .style("font-family", "Segoe UI")
    .style("line-height", "18px")
    .style("fill", "black")
    .style("font-weight", "400")
    .style("font-style", "normal");

  if (showLegendValues) {
    legend
      .append("text")
      .attr("x", 25)
      .attr("y", 100)
      .text((d) => `${d3Format.format(",.0f")(d.value)}`)
      .style("font-size", "28px")
      .style("font-family", "Segoe UI")
      .style("line-height", "34px")
      .style("fill", "#24272b")
      .style("font-style", "normal")
      .style("font-weight", "600");
  }
}

export function drawText(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  text: string,
  x: number,
  y: number,
  styles: TextStyle,
) {
  const alignItems = styles.alignItems ? styles.alignItems : "";
  const textAlign = styles.textAlign ? styles.textAlign : "";
  const textAnchor = styles.textAnchor ? styles.textAnchor : "";

  svg
    .append("text")
    .attr("x", x)
    .attr("y", y)
    .text(text)
    .raise()
    .style("font-family", styles.fontFamily)
    .style("font-style", styles.fontStyle)
    .style("font-weight", styles.fontWeight)
    .style("font-size", styles.fontSize)
    .style("line-height", styles.lineHeight)
    .style("display", styles.display)
    .style("align-items", alignItems)
    .style("text-align", textAlign)
    .style("fill", styles.color)
    .style("text-anchor", textAnchor);
}

export function drawArrow(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  color: string,
  id: string,
) {
  const markerDim = 3;
  const refX = 2;
  const refY = 1.5;
  const arrowPoints = [
    [0, 0],
    [0, markerDim],
    [markerDim, refY],
  ];

  svg
    .append("marker")
    .attr("id", id)
    .attr("refX", refX)
    .attr("refY", refY)
    .attr("markerWidth", markerDim)
    .attr("markerHeight", markerDim)
    .style("fill", color)
    .attr("orient", "auto-start-reverse")
    .append("path")
    .attr("d", d3.line()(arrowPoints as any));
}

export function drawLineTick(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  width: number,
  height: number,
  color: string,
  id: string,
) {
  const refX = 1;
  const refY = height / 2;

  const markerPoints = [
    [0, 0],
    [0, height],
    [width, height],
    [width, 0],
  ];

  svg
    .append("defs")
    .append("marker")
    .attr("id", id)
    .attr("refX", refX)
    .attr("refY", refY)
    .attr("markerHeight", height)
    .attr("markerWidth", width)
    .style("fill", color)
    .style("stroke", color)
    .attr("orient", "auto-start-reverse")
    .append("path")
    .attr("d", d3.line()(markerPoints as any));
}

export function drawCircleMarkers(
  svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
  width: number,
  height: number,
  color: string,
  radius: number,
  refX: number,
  refY: number,
  cx: number,
  cy: number,
  id: string,
) {
  svg
    .append("defs")
    .append("marker")
    .attr("id", id)
    .attr("refX", refX)
    .attr("refY", refY)
    .attr("markerWidth", width)
    .attr("markerHeight", height)
    .append("circle")
    .attr("cx", cx)
    .attr("cy", cy)
    .attr("r", radius)
    .style("fill", color);
}
