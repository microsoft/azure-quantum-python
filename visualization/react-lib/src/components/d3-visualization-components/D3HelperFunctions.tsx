import * as d3 from "d3";
import "./CSS/Shared.css";
import * as d3Format from 'd3-format';

export type LegendData = {
    title: string;
    legendTitle: string;
    value: number;
};

export function drawEllipses(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    cx: number,
    cy: number,
    spaceBetween: number,
    radius: number,
    fillColor: string
) {
    svg
        .append("circle")
        .attr("cx", cx)
        .attr("cy", cy)
        .attr("fill", fillColor)
        .attr("r", radius);

    svg
        .append("circle")
        .attr("cx", cx + spaceBetween)
        .attr("cy", cy)
        .attr("fill", fillColor)
        .attr("r", radius);

    svg
        .append("circle")
        .attr("cx", cx + spaceBetween * 2)
        .attr("cy", cy)
        .attr("fill", fillColor)
        .attr("r", radius);
}

export function drawLine(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    linePoints: number[][],
    className: string,
    id: string,
    strokeWidth: string,
    markerStart: string,
    markerEnd: string,
    fillColor: string,
    strokeColor: string,
    isDashed: boolean
) {
    // Create line generator
    const lineGenerator = d3.line();
    if (isDashed) {
        svg
            .append("path")
            .attr("class", className)
            .attr("id", id)
            .attr("d", lineGenerator(linePoints as any))
            .attr("stroke-width", strokeWidth)
            .attr("marker-start", markerStart)
            .attr("marker-end", markerEnd)
            .style("fill", fillColor)
            .style("stroke-dasharray", "3,3")
            .style("stroke", strokeColor);
    } else {
        svg
            .append("path")
            .attr("class", className)
            .attr("id", id)
            .attr("d", lineGenerator(linePoints as any))
            .attr("stroke-width", strokeWidth)
            .attr("marker-start", markerStart)
            .attr("marker-end", markerEnd)
            .style("fill", fillColor)
            .style("stroke", strokeColor)
            .lower();
    }
}

export function drawLegend(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    legendData: LegendData[],
    midpoint: number,
    chartBottomY: number,
    chartStartX: number,
    legendColor: d3.ScaleOrdinal<string, unknown, never>,
    showLegendValues: boolean,
    useTitleForColor: boolean
) {
    const legend = svg
        .selectAll(".legend")
        .data(legendData)
        .enter()
        .append("g")
        .attr("class", "legend")
        .attr(
            "transform",
            (d, i) => `translate(${midpoint * i + chartStartX}, ${chartBottomY})`
        );

    if (useTitleForColor) {
        legend
            .append("rect")
            .attr("width", 20)
            .attr("height", 20)
            .attr("x", 0)
            .attr("y", 50)
            .style("fill", (d, i) => legendColor(d.title) as string);
    }
    else {
        legend
            .append("rect")
            .attr("width", 20)
            .attr("height", 20)
            .attr("x", 0)
            .attr("y", 50)
            .style("fill", (d, i) => legendColor(d.legendTitle) as string);
    }

    legend
        .append("text")
        .attr("x", 25)
        .attr("y", 60)
        .text((d) => `${d.title}`)
        .attr("class", "legend");

    legend
        .append("text")
        .attr("x", 25)
        .attr("y", 75)
        .text((d) => `${d.legendTitle}`)
        .attr("class", "legend-maintext");

    if (showLegendValues) {
        legend
            .append("text")
            .attr("x", 25)
            .attr("y", 100)
            .text((d) => `${d3Format.format(',.0f')(d.value)}`)
            .attr("class", "legend-values");
    }

}

export function drawText(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    text: string,
    x: number,
    y: number,
    className: string
) {
    svg
        .append("text")
        .attr("x", x)
        .attr("y", y)
        .text(text)
        .raise()
        .attr("class", className);
}

export function drawArrow(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    color: string,
    id: string
) {
    const markerDim = 3;
    const refX = 2;
    const refY = 1.5;
    const arrowPoints = [
        [0, 0],
        [0, markerDim],
        [markerDim, refY],
    ];

    svg
        .append("marker")
        .attr("id", id)
        .attr("refX", refX)
        .attr("refY", refY)
        .attr("markerWidth", markerDim)
        .attr("markerHeight", markerDim)
        .style("fill", color)
        .attr("orient", "auto-start-reverse")
        .append("path")
        .attr("d", d3.line()(arrowPoints as any));
}

export function drawLineTick(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    width: number,
    height: number,
    color: string,
    id: string
) {
    const refX = 1;
    const refY = height / 2;

    const markerPoints = [
        [0, 0],
        [0, height],
        [width, height],
        [width, 0],
    ];

    svg
        .append("defs")
        .append("marker")
        .attr("id", id)
        .attr("refX", refX)
        .attr("refY", refY)
        .attr("markerHeight", height)
        .attr("markerWidth", width)
        .style("fill", color)
        .style("stroke", color)
        .attr("orient", "auto-start-reverse")
        .append("path")
        .attr("d", d3.line()(markerPoints as any));
}

export function drawCircleMarkers(
    svg: d3.Selection<d3.BaseType, unknown, HTMLElement, any>,
    width: number,
    height: number,
    color: string,
    radius: number,
    refX: number,
    refY: number,
    cx: number,
    cy: number,
    id: string
) {
    svg
        .append("defs")
        .append("marker")
        .attr("id", id)
        .attr("refX", refX)
        .attr("refY", refY)
        .attr("markerWidth", width)
        .attr("markerHeight", height)
        .append("circle")
        .attr("cx", cx)
        .attr("cy", cy)
        .attr("r", radius)
        .style("fill", color);
}
