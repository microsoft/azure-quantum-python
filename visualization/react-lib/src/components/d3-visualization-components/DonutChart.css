text.tooltip {
    background-color: pink;
    padding: 10px;
    box-shadow: 4px 4px 10px rgba(0, 0, 0, 0.4);
    border : 1px solid black;
    color: blue;
}

.donut-chart{
    position: absolute;
    font-family: 'Segoe UI';
    font-style: normal;
    font-weight: 600;
    font-size: 24px;
    line-height: 32px;
    
    align-items:center;
    color:  #24272B;
    margin:auto;

    display: inline-block;
    align-items: center;

    top: 30%;
    bottom: 0%;
    left: 25%;

}

