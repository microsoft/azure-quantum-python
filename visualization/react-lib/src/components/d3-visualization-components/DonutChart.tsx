/*------------------------------------
  Copyright (c) Microsoft Corporation.
  Licensed under the MIT License.
  All rights reserved.
------------------------------------ */
import * as React from "react";
import * as d3 from "d3";
import * as d3Format from "d3-format";
import { PieArcDatum } from "d3-shape";

import * as d3Helper from "./D3HelperFunctions";
import { TextStyle } from "./D3HelperFunctions";

export type DonutChartProps = {
  data: d3Helper.LegendData[];
  width: number;
  height: number;
  innerRadius: number;
  outerRadius: number;
};

/* Define styles */
const titleStyle: TextStyle = {
  fontFamily: "Segoe UI",
  fontStyle: "normal",
  fontWeight: "600",
  fontSize: "35",
  lineHeight: "47",
  display: "flex",
  alignItems: "center",
  textAlign: "center",
  color: "#201f1e",
  textAnchor: "middle",
};

const donutMiddleTitleStyle: TextStyle = {
  fontSize: "18",
  color: "#323130",
  fontWeight: "400",
  fontStyle: "normal",
  textAnchor: "middle",
  fontFamily: "Segoe UI",
  lineHeight: "21",
  alignItems: "center",
  textAlign: "center",
  display: "flex",
};

const donutMiddleTextStyle: TextStyle = {
  fontSize: "55",
  color: "#323130",
  fontWeight: "400",
  fontStyle: "normal",
  textAnchor: "middle",
  fontFamily: "Segoe UI",
  lineHeight: "73",
  alignItems: "center",
  textAlign: "center",
  display: "flex",
};

function DonutChart({
  data,
  width,
  height,
  innerRadius,
  outerRadius,
}: DonutChartProps) {
  React.useEffect(() => {
    /* ------------------------------------------------------------ Set up and define constants ------------------------------------------------------------  */
    const svg = d3.select("#donutchart");
    svg.selectAll("*").remove();

    /*------------------------------  Define chart dimensions ------------------------------  */
    const innerRadiusHover = innerRadius;
    const outerRadiusHover = outerRadius + 25;
    const donutMiddleTitle = "Total physical qubits";
    const padAngle = 0.01;

    const translationValX: number = width / 4;
    const translationValY: number = height / 4;

    /* ------------------------------  Define chart styling constants ------------------------------ */
    const chartOpacity = 0.75;
    const chartHoverOpacity = 1;
    const colorArray = ["#1a5d8c", "#8c1a5c", "#aebac0", "#323130"];

    /*------------------------------  Define color ranges  ------------------------------  */
    const algorithmRunTimeColor = colorArray[1];
    const tfactoryLineColor = colorArray[0];

    const chartColor = d3
      .scaleOrdinal()
      .domain(
        d3.extent(data, (d) => {
          return d.legendTitle;
        }) as unknown as string,
      )
      .range([algorithmRunTimeColor, tfactoryLineColor]);

    /* ------------------------------------------------------------ Begin draw chart ------------------------------------------------------------  */

    /*------------------------------  Create pie and arc generators  ------------------------------  */
    const pieGenerator = d3
      .pie<d3Helper.LegendData>()
      .padAngle(padAngle)
      .value((d) => d.value)
      .sort(null);

    const arcGenerator = d3
      .arc<PieArcDatum<d3Helper.LegendData>>()
      .innerRadius(innerRadius)
      .outerRadius(outerRadius);

    const pieData = pieGenerator(data);

    const arcs = svg
      .selectAll("g")
      .data(pieData)
      .enter()
      .append("g")
      .attr("class", "arc")
      .attr(
        "transform",
        `translate(${innerRadius + translationValX},${
          innerRadius + translationValY
        })`,
      );

    /*------------------------------  Fill donut chart and apply hover  ------------------------------  */
    arcs
      .append("path")
      .attr("d", arcGenerator)
      .attr("fill", (d) => {
        return chartColor(d.data.legendTitle) as string;
      })
      .style("opacity", chartOpacity)
      .on("mouseover", (d) => {
        const arcHover = d3
          .arc<PieArcDatum<d3Helper.LegendData>>()
          .innerRadius(innerRadiusHover)
          .outerRadius(outerRadiusHover);
        d3.select(d.target).attr("d", arcHover as any);
        d3.select(d.target).style("opacity", chartHoverOpacity);
      })
      .on("mouseout", (d) => {
        d3.select(d.target).attr("d", arcGenerator as any);
        d3.select(d.target).style("opacity", chartOpacity);
      });

    // Add tooltips
    arcs
      .append("title")
      .text(
        (d) =>
          `${d.data.title} ${d.data.legendTitle} : ${d3Format.format(",.0f")(
            d.value,
          )}`,
      );

    /*------------------------------  Draw Legend ------------------------------  */
    const legendY = translationValY + outerRadius * 2 + 10;
    const midpoint = outerRadius + 10;
    d3Helper.drawLegend(
      svg,
      data,
      midpoint,
      legendY,
      translationValX,
      chartColor,
      true,
      false,
    );

    /*------------------------------  Add text and titles  ------------------------------  */

    // Add chart title
    d3Helper.drawText(
      svg,
      "Space diagram",
      innerRadius + translationValX,
      translationValY - innerRadius,
      titleStyle,
    );

    // Add middle text
    const totalQubits = d3.sum(data, (d) => d.value);
    const totalQubitsStr = d3Format.format(",.0f")(totalQubits);

    d3Helper.drawText(
      svg,
      donutMiddleTitle,
      innerRadius + translationValX,
      translationValY + innerRadius - 25,
      donutMiddleTitleStyle,
    );
    d3Helper.drawText(
      svg,
      totalQubitsStr,
      innerRadius + translationValX,
      translationValY + innerRadius + 25,
      donutMiddleTextStyle,
    );
  }, [data, innerRadius, outerRadius]);

  return (
    <div
      style={{
        justifyContent: "center",
        alignItems: "center",
      }}
    >
      <svg id="donutchart" width={width} height={height}></svg>
    </div>
  );
}

export default DonutChart;
