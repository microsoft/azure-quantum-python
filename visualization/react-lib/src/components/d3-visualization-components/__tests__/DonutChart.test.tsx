/*------------------------------------
  Copyright (c) Microsoft Corporation.
  Licensed under the MIT License.
  All rights reserved.
------------------------------------ */
import React from "react";
import { create } from "react-test-renderer";

import { LegendData } from "../D3HelperFunctions";
import DonutChart, { DonutChartProps } from "../DonutChart";

describe("Donut chart tests", () => {
  it("Verify Donut Chart", () => {
    const testData: LegendData[] = [
      {
        legendTitle: "Logical qubits",
        title: "Logical qubits",
        value: 50,
      },
      {
        legendTitle: "Physical qubits",
        title: "Physical qubits",
        value: 200,
      },
    ];

    const donutProps: DonutChartProps = {
      data: testData,
      width: 1000,
      height: 1000,
      innerRadius: 100,
      outerRadius: 200,
    };

    const component = create(<DonutChart {...donutProps}></DonutChart>);
    expect(component.toJSON()).toMatchSnapshot("DonutChart");
  });
});
