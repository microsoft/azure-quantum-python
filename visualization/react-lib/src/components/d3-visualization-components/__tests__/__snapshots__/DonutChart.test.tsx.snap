// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Donut chart tests Verify Donut Chart: DonutChart 1`] = `
<div
  style={
    {
      "alignItems": "center",
      "justifyContent": "center",
    }
  }
>
  <svg
    height={1000}
    id="donutchart"
    width={1000}
  />
</div>
`;
